c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -364,6 +365,68 @@ struct PrivatePool {
   BlockPool small_blocks;
 };
 
+std::shared_ptr<BlockState> constructBlockState(Block* block) {
+  TORCH_CHECK(
+      block->event_count == 0,
+      "Events should have synchronized when checkpointing block");
+  std::shared_ptr<BlockState> bs = std::make_shared<BlockState>();
+  bs->stream = block->stream;
+  bs->stream_uses = block->stream_uses;
+  bs->size = block->size;
+  bs->ptr = block->ptr;
+  bs->allocated = block->allocated;
+  bs->gc_count = block->gc_count;
+  return bs;
+};
+
+BlockPoolState constructBlockPoolState(
+    BlockPool* block_pool,
+    MempoolId_t pool_id,
+    const std::vector<Block*>& pool_blocks) {
+  BlockPoolState bps;
+  bps.is_small = block_pool->is_small;
+  bps.owner_id = bps.owner_id;
+
+  std::unordered_map<Block*, BlockState*> mappings;
+
+  for (Block* block : pool_blocks) {
+    if (block->pool != block_pool) {
+      continue;
+    }
+
+    bps.blocks.emplace_back(constructBlockState(block));
+    mappings[block] = bps.blocks.back().get();
+
+    if (block->prev && mappings.count(block->prev)) {
+      BlockState* prev = mappings[block->prev];
+      prev->next = bps.blocks.back().get();
+      bps.blocks.back()->prev = prev;
+    }
+
+    if (block->next && mappings.count(block->next)) {
```

Comment:
How do you know that the block is in the mapping already

Reply:
Not sure I follow the question. But I can simplify the implementation without a map and serialize by just following Segment pointers. Will update